File: android/app/src/main/java/com/sefermormon/standardworks/SearchIndex.kt
package com.sefermormon.standardworks

import android.content.Context
import android.util.JsonReader
import android.util.JsonToken
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import java.io.StringReader
import java.util.concurrent.Executors

/**
 * SEARCH ACROSS ALL SIX VOLUMES, NATIVE — the twin of SearchIndex.swift.
 *
 * The site precomputes one cross-volume verse index for its home page
 * (search_index.js: every verse as [deep-link, reference, consonantal Hebrew,
 * English]). Read once off the main thread, streamed rather than parsed
 * whole, and searched the way the home page searches: Hebrew with or without
 * nikkud, English case-blind, hits in the canon's order, 200 per volume.
 */
class SearchIndex(private val context: Context) {
    class Row(val volume: String, val link: String, val ref: String, val heb: String, val eng: String)
    class Hit(val row: Row, val volumeName: String, val page: String, val snippet: String) {
        val id get() = row.volume + "#" + row.link
        val path get() = "$page#${row.link}"
    }

    var volumes: List<String> = emptyList(); private set
    private var names: Map<String, String> = emptyMap()
    private var pages: Map<String, String> = emptyMap()
    private var rows: Map<String, List<Row>> = emptyMap()
    private var lowered: Map<String, List<String>> = emptyMap()
    var loaded by mutableStateOf(false); private set
    /** Book name → its place in the canon (WebShell sets it from the registry). */
    @Volatile var bookOrder: Map<String, Int> = emptyMap()

    init { Executors.newSingleThreadExecutor().execute { load() } }

    private fun load() {
        val text = try { context.assets.open("www/search_index.js").bufferedReader().use { it.readText() } } catch (e: Exception) { return }
        val start = text.indexOf('{'); if (start < 0) return
        val end = text.lastIndexOf('}'); if (end < start) return
        val vols = ArrayList<String>(); val nm = HashMap<String, String>(); val pg = HashMap<String, String>()
        val rw = HashMap<String, List<Row>>(); val low = HashMap<String, List<String>>()
        try {
            JsonReader(StringReader(text.substring(start, end + 1))).use { r ->
                r.beginObject()
                while (r.hasNext()) {
                    when (r.nextName()) {
                        "vols" -> { r.beginArray(); while (r.hasNext()) vols.add(r.nextString()); r.endArray() }
                        "names" -> { r.beginObject(); while (r.hasNext()) nm[r.nextName()] = r.nextString(); r.endObject() }
                        "pages" -> { r.beginObject(); while (r.hasNext()) pg[r.nextName()] = r.nextString(); r.endObject() }
                        "rows" -> {
                            r.beginObject()
                            while (r.hasNext()) {
                                val v = r.nextName()
                                val list = ArrayList<Row>()
                                r.beginArray()
                                while (r.hasNext()) {
                                    r.beginArray()
                                    val cells = ArrayList<String>(4)
                                    while (r.hasNext()) { if (r.peek() == JsonToken.NULL) { r.nextNull(); cells.add("") } else cells.add(r.nextString()) }
                                    r.endArray()
                                    if (cells.size >= 4) list.add(Row(v, cells[0], cells[1], cells[2], cells[3]))
                                }
                                r.endArray()
                                rw[v] = list
                                low[v] = list.map { it.eng.lowercase() }
                            }
                            r.endObject()
                        }
                        else -> r.skipValue()
                    }
                }
                r.endObject()
            }
        } catch (e: Exception) { return }
        volumes = vols; names = nm; pages = pg; rows = rw; lowered = low
        loaded = true
    }

    /** Hits in the site's volume order, at most `perVolume` from each. */
    fun find(query: String, perVolume: Int = 200): List<Hit> {
        val q = query.trim()
        if (!loaded || q.length < 2) return emptyList()
        val isHeb = hasHebrew(q)
        val nq = if (isHeb) normHeb(q) else q.lowercase()
        if (nq.isEmpty()) return emptyList()
        val hits = ArrayList<Hit>()
        val order = bookOrder
        for (v in volumes) {
            val list = rows[v] ?: continue; val lowList = lowered[v] ?: continue
            val found = ArrayList<Triple<Int, Int, Pair<Int, Hit>>>()
            for ((i, r) in list.withIndex()) {
                val hay = if (isHeb) r.heb else lowList[i]
                val at = hay.indexOf(nq); if (at < 0) continue
                val text = if (isHeb) r.heb else r.eng
                val (book, ch, vs) = place(r.ref)
                found.add(Triple(order[book] ?: Int.MAX_VALUE, ch, Pair(vs, Hit(r, names[v] ?: v, pages[v] ?: "", snippet(text, at, nq.length)))))
            }
            found.sortWith(compareBy({ it.first }, { it.second }, { it.third.first }))
            for (f in found.take(perVolume)) hits.add(f.third.second)
        }
        return hits
    }

    companion object {
        private val points = Regex("[\\u0591-\\u05C7]")
        private val finals = mapOf('ך' to 'כ', 'ם' to 'מ', 'ן' to 'נ', 'ץ' to 'צ', 'ף' to 'פ')

        fun hasHebrew(s: String) = s.any { it in '֐'..'׿' }

        fun normHeb(s: String): String {
            var t = s.replace("׃", "").replace("[", "").replace("]", "")
            t = points.replace(t, "")
            t = t.replace('־', ' ')
            t = t.map { finals[it] ?: it }.joinToString("")
            return t.split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ")
        }

        /** "1 Kings 3:3" → ("1 Kings", 3, 3) */
        fun place(ref: String): Triple<String, Int, Int> {
            val sp = ref.lastIndexOf(' '); if (sp < 0) return Triple(ref, 0, 0)
            val cv = ref.substring(sp + 1).split(':')
            return Triple(ref.substring(0, sp), cv.getOrNull(0)?.toIntOrNull() ?: 0, cv.getOrNull(1)?.toIntOrNull() ?: 0)
        }

        private fun snippet(text: String, start: Int, len: Int): String {
            if (text.isEmpty()) return text
            val lo = maxOf(0, start - 40); val hi = minOf(text.length, start + len + 90)
            if (lo >= hi) return text
            return (if (lo > 0) "…" else "") + text.substring(lo, hi) + (if (hi < text.length) "…" else "")
        }
    }
}

File: StandardWorks/SearchIndex.swift
import Foundation
import JavaScriptCore

/// SEARCH ACROSS ALL SIX VOLUMES, NATIVE.
///
/// The site precomputes one cross-volume verse index for its home page
/// (tools/build_search_index.js → search_index.js: every verse as
/// [deep-link, reference, consonantal Hebrew, English]). The reader pages
/// only index the volume they hold, which is why the Search tab first showed
/// two Old Testament verses for "charity" and none of Moroni 7. This reads
/// the home page's file instead — the whole canon — once, off the main
/// thread, evaluated in JavaScriptCore exactly as the page would evaluate it,
/// and searches it the way the home page does: Hebrew with or without nikkud
/// (points stripped, finals folded, maqqef a space), English case-blind.
final class SearchIndex: ObservableObject {
    struct Row {
        let volume: String
        let link: String       // the reader's deep-link hash
        let ref: String        // "3 Nephi 17:4"
        let heb: String        // consonantal
        let eng: String
    }
    struct Hit: Identifiable {
        let row: Row
        let volumeName: String
        let page: String
        let snippet: String
        var id: String { row.volume + "#" + row.link }
        var path: String { page + "#" + row.link }
    }

    private(set) var volumes: [String] = []
    private var names: [String: String] = [:]
    private var pages: [String: String] = [:]
    private var rows: [String: [Row]] = [:]
    private var lowered: [String: [String]] = [:]
    @Published private(set) var loaded = false
    /// Book name → its place in the canon, from the site's registry (WebShell
    /// sets it), so hits read Genesis before 1 Kings and not in file order.
    var bookOrder: [String: Int] = [:]
    private let queue = DispatchQueue(label: "search-index", qos: .userInitiated)

    init(www: URL) {
        queue.async { [weak self] in self?.load(www: www) }
    }

    private func load(www: URL) {
        guard let js = try? String(contentsOf: www.appendingPathComponent("search_index.js"), encoding: .utf8),
              let ctx = JSContext() else { return }
        ctx.evaluateScript("var window = {};")
        ctx.evaluateScript(js)
        guard let json = ctx.evaluateScript("JSON.stringify(window.SW_SEARCH_INDEX)")?.toString(),
              let data = json.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }
        let vols = obj["vols"] as? [String] ?? []
        let nm = obj["names"] as? [String: String] ?? [:]
        let pg = obj["pages"] as? [String: String] ?? [:]
        let rw = obj["rows"] as? [String: [[Any]]] ?? [:]
        var out: [String: [Row]] = [:], low: [String: [String]] = [:]
        for v in vols {
            let list = (rw[v] ?? []).compactMap { r -> Row? in
                guard r.count >= 4 else { return nil }
                return Row(volume: v, link: r[0] as? String ?? "", ref: r[1] as? String ?? "",
                           heb: r[2] as? String ?? "", eng: r[3] as? String ?? "")
            }
            out[v] = list
            low[v] = list.map { $0.eng.lowercased() }
        }
        DispatchQueue.main.async {
            self.volumes = vols; self.names = nm; self.pages = pg; self.rows = out; self.lowered = low
            self.loaded = true
        }
    }

    // MARK: - matching, the home page's rules

    private static let points = try! NSRegularExpression(pattern: "[\\u0591-\\u05C7]")
    private static let finals: [Character: Character] = ["\u{05DA}": "\u{05DB}", "\u{05DD}": "\u{05DE}", "\u{05DF}": "\u{05E0}", "\u{05E5}": "\u{05E6}", "\u{05E3}": "\u{05E4}"]

    static func hasHebrew(_ s: String) -> Bool { s.unicodeScalars.contains { (0x0590...0x05FF).contains($0.value) } }

    static func normHeb(_ s: String) -> String {
        var t = s.replacingOccurrences(of: "\u{05C3}", with: "").replacingOccurrences(of: "[", with: "").replacingOccurrences(of: "]", with: "")
        t = points.stringByReplacingMatches(in: t, range: NSRange(t.startIndex..., in: t), withTemplate: "")
        t = t.replacingOccurrences(of: "\u{05BE}", with: " ")
        t = String(t.map { finals[$0] ?? $0 })
        return t.split(whereSeparator: { $0.isWhitespace }).joined(separator: " ")
    }

    /// Hits in the site's volume order, at most `perVolume` from each.
    func find(_ query: String, perVolume: Int = 200) -> [Hit] {
        let q = query.trimmingCharacters(in: .whitespaces)
        guard loaded, q.count >= 2 else { return [] }
        let isHeb = Self.hasHebrew(q)
        let nq = isHeb ? Self.normHeb(q) : q.lowercased()
        guard !nq.isEmpty else { return [] }
        var hits: [Hit] = []
        for v in volumes {
            let list = rows[v] ?? [], low = lowered[v] ?? []
            var found: [(Int, Int, Int, Hit)] = []
            for (i, r) in list.enumerated() {
                let hay = isHeb ? r.heb : low[i]
                guard let range = hay.range(of: nq) else { continue }
                let text = isHeb ? r.heb : r.eng
                let (book, ch, vs) = Self.place(r.ref)
                found.append((bookOrder[book] ?? Int.max, ch, vs,
                              Hit(row: r, volumeName: names[v] ?? v, page: pages[v] ?? "",
                                  snippet: Self.snippet(text, around: range, in: hay))))
            }
            // The canon's order, then chapter and verse; the cap after the sort
            // so Genesis is never crowded out by 1 Chronicles.
            found.sort { ($0.0, $0.1, $0.2) < ($1.0, $1.1, $1.2) }
            hits.append(contentsOf: found.prefix(perVolume).map { $0.3 })
        }
        return hits
    }

    /// "1 Kings 3:3" → ("1 Kings", 3, 3)
    static func place(_ ref: String) -> (String, Int, Int) {
        guard let sp = ref.lastIndex(of: " ") else { return (ref, 0, 0) }
        let book = String(ref[..<sp]), cv = ref[ref.index(after: sp)...].split(separator: ":")
        return (book, Int(cv.first ?? "") ?? 0, Int(cv.count > 1 ? cv[1] : "") ?? 0)
    }

    /// A window of the verse around the match; the offsets of the lowered
    /// haystack are the offsets of the text it was lowered from.
    private static func snippet(_ text: String, around range: Range<String.Index>, in hay: String) -> String {
        let start = hay.distance(from: hay.startIndex, to: range.lowerBound)
        let len = hay.distance(from: range.lowerBound, to: range.upperBound)
        let chars = Array(text)
        guard !chars.isEmpty else { return text }
        let lo = max(0, start - 40), hi = min(chars.count, start + len + 90)
        guard lo < hi, hi <= chars.count else { return text }
        return (lo > 0 ? "\u{2026}" : "") + String(chars[lo..<hi]) + (hi < chars.count ? "\u{2026}" : "")
    }
}
